website/content/docs/auth/jwt/oidc_providers.mdx

```diff
@@ -253,6 +253,220 @@ vault write auth/oidc/role/your_default_role \
+   ```
+
+1. Find the issuer URL of the cluster.
+
+   ```bash
+   kubectl proxy &
+   ISSUER="$(curl --fail --silent --show-error 127.0.0.1:8001/.well-known/openid-configuration | jq -r '.issuer')"
+
+   # Kill the background proxy process when you're done
+   kill %%
+   ```
+
+1. Enable and configure JWT auth in Vault.
+
+   1. If Vault is running in Kubernetes:
+
+      ```bash
+      kubectl exec vault-0 -- vault auth enable jwt
+      kubectl exec vault-0 -- vault write auth/jwt/config \
+         oidc_discovery_url=https://kubernetes.default.svc.cluster.local \
+         oidc_discovery_ca_pem=@/var/run/secrets/kubernetes.io/serviceaccount/ca.crt
+      ```
+
+   1. Alternatively, if Vault is _not_ running in Kubernetes:
+
+      -> **Note:** When Vault is outside the cluster, the `$ISSUER` endpoint below may
+         or may not be reachable. If not, you can configure JWT auth using
+         [`jwt_validation_pubkeys`](#using-jwt-validation-public-keys) instead.
+
+      ```bash
+      vault auth enable jwt
+      vault write auth/jwt/config oidc_discovery_url="${ISSUER}"
+      ```
+
+1. Configure a role and log in as detailed [below](#creating-a-role-and-logging-in).
+
+[k8s-sa-issuer-discovery]: https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-issuer-discovery
+
+### Using JWT validation public keys
+
+This method can be useful if Kubernetes' API is not reachable from Vault or if
+you would like a single JWT auth mount to service multiple Kubernetes clusters
+by chaining their public signing keys.
+
+Kubernetes cluster requirements:
+
+* [`ServiceAccountIssuerDiscovery`][k8s-sa-issuer-discovery] feature enabled.
+   * Present from 1.18, defaults to enabled from 1.20.
+   * This requirement can be avoided if you can access the Kubernetes master
+      nodes to read the public signing key directly from disk at
+      `/etc/kubernetes/pki/sa.pub`. In this case, you can skip the steps to
+      retrieve and then convert the key as it will already be in PEM format.
+* Must use short-lived service account tokens when logging in.
+   * Tokens mounted into pods default to short-lived from 1.21.
+
+Configuration steps:
+
+1. Fetch the service account signing public key from your cluster's JWKS URI.
+
+   ```bash
+   # 1. Find the issuer URL of the cluster.
+   kubectl proxy &
+   ISSUER="$(curl --fail --silent --show-error 127.0.0.1:8001/.well-known/openid-configuration | jq -r '.issuer')"
+
+   # 2. Query the jwks_uri specified in /.well-known/openid-configuration
+   # NB: You may need to run this from a pod within the cluster if the $ISSUER
+   # URL is not available outside the cluster.
+   curl "$(curl --fail --silent --show-error "${ISSUER}/.well-known/openid-configuration" | jq -r '.jwks_uri')"
+
+   # Kill the background proxy process when you're done
+   kill %%
+   ```
+
+1. Convert the keys from JWK format to PEM. You can use a CLI tool or an online
```

Comment:
I wonder if it'd be possible to suggest configuring the JWT auth method using [jwks_url](https://www.vaultproject.io/api-docs/auth/jwt#jwks_url) instead of [jwt_validation_pubkeys](https://www.vaultproject.io/api-docs/auth/jwt#jwt_validation_pubkeys). That is a simpler setup and would eliminate the step of converting JWKS -> PEM here. That configuration also accounts for the scenario where a public key in the JWKS is rotated, where `jwt_validation_pubkeys` is static (i.e., only uses the public keys you give it). I'm not sure if k8s rotates the keys they use to sign/verify JWTs.

It looks like the JWKS is published publicly from the k8s docs. I don't have much context from the k8s side of things, but it _seems_ possible to configure JWT auth with just the `jwks_url`. 
> When enabled, the Kubernetes API server provides an OpenID Provider Configuration document at /.well-known/openid-configuration and the associated JSON Web Key Set (JWKS) at /openid/v1/jwks.

Reply:
My thinking was that `jwt_validation_pubkeys` is definitely second choice to using the OIDC discovery URL for the same benefits you state for `jwks_url` (hence discovery coming first here), but it has some nice benefits too like the fact you can configure it to trust multiple clusters and can be permanently network-partitioned from the cluster.

As a layman, using `jwks_url` seemed basically equivalent to `oidc_discovery_url` in terms of the benefits and drawbacks. Does `jwks_url` have any advantages over `oidc_discovery_url`? You're right that we definitely can configure JWT auth with `jwks_url` for Kubernetes, but I don't think there are any (common?) configurations where Kubernetes could support `jwks_url` but couldn't support `oidc_discovery_url`. If there are distinct advantages it has though, then I'm definitely keen to add `jwks_url` as a third option.

Reply:
You're right that `jwks_url` doesn't have any distinct advantages over `oidc_discovery_url` other than the issuer check limitation that we discussed outside of this PR. That's the only reason I brought it up. Looks like you have some good guidance in the documentation for `oidc_discovery_url`, so keeping `jwt_validation_pubkeys` documented here sounds good to me 👍